### ISSUE TITLE
DOC: Single Document For Code Guidelines

### ISSUE THREAD

**Issue by Stockfoot** [pandas-dev/pandas#33851]
#### Location of the documentation

https://pandas.pydata.org/docs/development/code_style.html
https://pandas.pydata.org/pandas-docs/stable/development/contributing.html#code-standards

#### Documentation problem

Most of the documentation for the coding style guidelines is limited and in multiple locations. Much of the coding style is in a script.

#### Suggested fix for documentation

I propose to create a single document that contains all the coding guidelines.
This will allow new contributors to easily see all the required syntax and style for the project.
This also allows newer users to not have to guess against a script or have to keep submitting code over and over to find mistakes.

**Comment by simonjayhawkins** [comment]
> I propose to create a single document that contains all the coding guidelines.

perhaps the google style could be a good template or basis for outline http://google.github.io/styleguide/pyguide.html

personally i prefer a well structured reference style document to a informal prose on the subject. 

I think that the reason we have the two documents is that the contributing guidelines contains the descriptive introduction for first time contributors and the code style document contains more details. This second document was only recently introduced and adding the content and some migration of content from the contributing guidlines is WIP.

contributions and PRs welcome.

**Comment by Stockfoot** [comment]
I agree that the google style is a good template. I am currently in an Open Source course and would like to contribute to this.

I am a first time contributor and have never really done this so I have a question that I hope you can answer.

Is it useful/okay practice if I get the outline of the document and the information that already exists about the code style into the document or do I need to try and have a complete document for a pull request?

**Comment by simonjayhawkins** [comment]
> Is it useful/okay practice if I get the outline of the document and the information that already exists about the code style into the document or do I need to try and have a complete document for a pull request?

smaller more atomic PRs are easier to review and will likely get approved/merged quicker.

Ok to start small, any incremental improvements welcome., but probably better to not have a document with empty sections. so could use the google doc for inspiration but not actually have a document with an outline and missing content.

**Comment by Stockfoot** [comment]
@simonjayhawkins I have begun writing the document in markdown and saw in the documentation section that the project uses Sphinx. Should I switch over to Sphinx before proceeding further or is markdown okay?

**Comment by moaraccounts** [comment]
@Stockfoot I'm new to GitHub, learning first by contributing to documentation. Do you need/want any help?

**Comment by Stockfoot** [comment]
@moaraccounts sure that would be good. Like I said I have a markdown file going but I'm not sure if I need to switch to Sphinx.
I have pulled the coding guidelines from various places that was already presented. I think the next step is I need to pull unique requirements from the style script but I am not sure how to do that.

**Comment by moaraccounts** [comment]
@Stockfoot I've been looking at the script, trying to understand the requirements. What if you created the documentation in a similar manner to how the code_checks script is constructed? Like, have one section on Linting and pull in the different messages from the other scripts as subsections? Is that what you mean by pulling the requirements from the style script?
For example:
<Documentation Title> Linting
    <Subsection> Check for use of not concatenated strings (This is from code_checks.sh)
        <Example>Test Case for bare pytest raises (This whole example would be from validate_unwanted_patterns.py, lines 57-114)

**Comment by Stockfoot** [comment]
Yeah that is what I am getting at. Basically a style guide that would allow a first time user to read it and understand the formatting and requirements for how they should code, The script would act as a backup instead of a newer user having to submit over and over to see where they are formatting things wrong.
[pandas Coding Style Guide example.pdf](https://github.com/pandas-dev/pandas/files/4659392/pandas.Coding.Style.Guide.example.pdf)
Here is the markdown file I have going, I am trying to format it kind of like the python style guide

**Comment by sahilpatni95** [comment]
I am a new in contributor and have never really done this so I have a question that I hope you can answer.

**Comment by moaraccounts** [comment]
> @moaraccounts
> Yeah that is what I am getting at. Basically a style guide that would allow a first time user to read it and understand the formatting and requirements for how they should code, The script would act as a backup instead of a newer user having to submit over and over to see where they are formatting things wrong.

Awesome. I like the .pdf you have going. And I think you're fine using markdown (see [this](https://pandas.pydata.org/docs/development/contributing.html?highlight=sphinx#about-the-pandas-documentation)), but good to take a look at [this link](https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html) which points to the reStructuredText Primer just in case. Is there anything I can help with?

**Comment by Stockfoot** [comment]
@moaraccounts 
Thanks for the links! I will work on this some more this week and try submitting a pull request. If it gets accepted then it opens the door for everyone else to easily contribute to the document!

**Comment by Stockfoot** [comment]
@simonjayhawkins @moaraccounts 
I tried submitting a pull request for the .rst document but it failed. It is saying there is trailing white spaces after every single line on the CI/Check but when I pull it up in my editing program there are no trailing white spaces. I'm sure it is something simple but perhaps you can shed some light on what is happening. Thanks!

**Comment by moaraccounts** [comment]
@Stockfoot @simonjayhawkins I pulled Stockfoot's branch, compiled the docs and ran code_checks.sh. I'm seeing the trailing whitespace errors. However, not sure how to address it. I'm not using an IDE, just Terminal. Hmm. Actually, I'll try pull up the docs in Sublime Text or something and see if that gives me more info.

**Comment by Stockfoot** [comment]
I haven't submitted anything else because I am stuck with the trailing white space issue and am unsure of how to proceed

**Comment by moaraccounts** [comment]
@Stockfoot I hear that. I'm looking up ways to remove trailing whitespaces from .rst files. But I don't know how to push my changes back to your branch, ha. N00b to that. Any advice?

**Comment by Stockfoot** [comment]
@moaraccounts Fork my "Coding-Style-Guideline-Documentation" branch, make the changes, then do a pull request on that branch

**Comment by rajaryanbanka** [comment]
I would like to take this up.

**Comment by arneja-arnav** [comment]
This thread seems inactive. Is it alright if I assign myself to work on this issue?

**Comment by moaraccounts** [comment]
Certainly!

**Comment by harshvardhan2707** [comment]
take

**Comment by arneja-arnav** [comment]
Take

**Comment by james-wall** [comment]
Hi @arneja-arnav , is this issue available to be taken?

**Comment by phanr8541** [comment]
Hi, I'd like to tackle this issue if it is still available

**Comment by makroumi** [comment]
Hello, I am looking to formally take this on.

It looks like the last few attempts stalled due to two main issues: (1) the required conversion to reStructuredText/Sphinx, and (2) identifying content from the linting scripts.

To ensure a smooth, atomic PR, I propose the following phased approach, following @simonjayhawkins's advice for smaller PRs:

1.  **Initial PR (Phase 1):** Create the new `.rst` file and migrate *all* content currently present in the two existing documents (`code_style.html` and `contributing.html#code-standards`) into the new single document. This establishes the new file and fixes the immediate documentation duplication.
2.  **Follow-up PR (Phase 2):** Based on the structure from Phase 1, I will analyze the linting scripts (`code_checks.sh`, etc.) to extract unique rules not currently documented and add them to the new file.

I will start on **Phase 1 (Migration to .rst)** immediately.

@phanr8541 - Since I will be opening a PR shortly, please let me know if you are currently working on this, otherwise I will proceed with the claim.

**Comment by jbrockmendel** [comment]
closed by #46724
